File: python/src/lichen/timing/data_timing.py
# SPDX-License-Identifier: GPL-3.0-or-later
# SPDX-FileCopyrightText: The contributors to the LICHEN project
"""Data-traffic timing (spec 09 section 14.3).

Telemetry is 5-60 minutes. Heartbeat/keepalive is 30 minutes if no data
has been sent. Elapsed time is wrap-safe unsigned 64-bit.
"""

from __future__ import annotations

TELEMETRY_MIN_MS: int = 5 * 60 * 1000
TELEMETRY_MAX_MS: int = 60 * 60 * 1000
HEARTBEAT_MS: int = 30 * 60 * 1000
_U64: int = (1 << 64) - 1


def elapsed(now: int, then: int) -> int:
    """Wrap-safe ``now - then`` on the u64 tick ring."""
    if type(now) is not int or type(then) is not int:
        raise TypeError("ticks must be int")
    if now < 0 or then < 0:
        raise ValueError("ticks must be non-negative")
    return (now - then) & _U64


class TelemetryIntervalError(ValueError):
    """Interval outside 5-60 minutes."""


class TelemetryInterval:
    """Configured telemetry period in ``[5 min, 60 min]``."""

    __slots__ = ("_interval_ms",)

    def __init__(self, interval_ms: int) -> None:
        if type(interval_ms) is not int:
            raise TypeError("interval_ms must be int")
        if interval_ms < TELEMETRY_MIN_MS or interval_ms > TELEMETRY_MAX_MS:
            raise TelemetryIntervalError("telemetry interval must be 5-60 minutes")
        self._interval_ms = interval_ms

    @property
    def interval_ms(self) -> int:
        return self._interval_ms

    def due(self, now_ms: int, last_tx_ms: int | None) -> bool:
        if last_tx_ms is None:
            return True
        return elapsed(now_ms, last_tx_ms) >= self._interval_ms


class Heartbeat:
    """30-minute keepalive if no data has been sent."""

    __slots__ = ("_interval_ms",)

    def __init__(self) -> None:
        self._interval_ms = HEARTBEAT_MS

    def due(self, now_ms: int, last_data_tx_ms: int | None) -> bool:
        if last_data_tx_ms is None:
            return True
        return elapsed(now_ms, last_data_tx_ms) >= self._interval_ms

File: rust/lichen-link/src/data_timing.rs
//! Data-traffic timing (spec 09 section 14.3).
//!
//! Telemetry is 5-60 minutes. Heartbeat/keepalive is 30 minutes if no data
//! has been sent. Comparisons use wrapping unsigned elapsed time.

/// 5 minutes in milliseconds.
pub const TELEMETRY_MIN_MS: u64 = 5 * 60 * 1000;
/// 60 minutes in milliseconds.
pub const TELEMETRY_MAX_MS: u64 = 60 * 60 * 1000;
/// Heartbeat/keepalive interval (30 minutes).
pub const HEARTBEAT_MS: u64 = 30 * 60 * 1000;

/// Wrap-safe elapsed ticks: `now.wrapping_sub(then)`.
pub const fn elapsed(now: u64, then: u64) -> u64 {
    now.wrapping_sub(then)
}

/// Configured telemetry period.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TelemetryInterval {
    interval_ms: u64,
}

/// Invalid telemetry configuration.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TelemetryIntervalError {
    /// Below 5 minutes or above 60 minutes.
    OutOfRange,
}

impl TelemetryInterval {
    /// Construct a period in `[5 min, 60 min]`.
    pub const fn new(interval_ms: u64) -> Result<Self, TelemetryIntervalError> {
        if interval_ms < TELEMETRY_MIN_MS || interval_ms > TELEMETRY_MAX_MS {
            return Err(TelemetryIntervalError::OutOfRange);
        }
        Ok(Self { interval_ms })
    }

    /// Configured period in milliseconds.
    pub const fn interval_ms(self) -> u64 {
        self.interval_ms
    }

    /// True if no sample has been sent, or `interval_ms` has elapsed.
    pub const fn due(self, now_ms: u64, last_tx_ms: Option<u64>) -> bool {
        match last_tx_ms {
            None => true,
            Some(last) => elapsed(now_ms, last) >= self.interval_ms,
        }
    }
}

/// 30-minute heartbeat if no data has been sent.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Heartbeat {
    interval_ms: u64,
}

impl Heartbeat {
    /// Spec 14.3 keepalive interval.
    pub const fn new() -> Self {
        Self {
            interval_ms: HEARTBEAT_MS,
        }
    }

    /// True if no data has gone out, or 30 minutes have elapsed since last TX.
    pub const fn due(self, now_ms: u64, last_data_tx_ms: Option<u64>) -> bool {
        match last_data_tx_ms {
            None => true,
            Some(last) => elapsed(now_ms, last) >= self.interval_ms,
        }
    }
}

impl Default for Heartbeat {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn telemetry_rejects_out_of_range() {
        assert_eq!(
            TelemetryInterval::new(TELEMETRY_MIN_MS - 1),
            Err(TelemetryIntervalError::OutOfRange)
        );
        assert_eq!(
            TelemetryInterval::new(TELEMETRY_MAX_MS + 1),
            Err(TelemetryIntervalError::OutOfRange)
        );
        assert!(TelemetryInterval::new(TELEMETRY_MIN_MS).is_ok());
        assert!(TelemetryInterval::new(TELEMETRY_MAX_MS).is_ok());
    }

    #[test]
    fn telemetry_due_after_interval() {
        let period = TelemetryInterval::new(TELEMETRY_MIN_MS).unwrap();
        assert!(period.due(0, None));
        assert!(!period.due(TELEMETRY_MIN_MS - 1, Some(0)));
        assert!(period.due(TELEMETRY_MIN_MS, Some(0)));
    }

    #[test]
    fn heartbeat_due_after_30_min() {
        let hb = Heartbeat::new();
        assert!(hb.due(0, None));
        assert!(!hb.due(HEARTBEAT_MS - 1, Some(0)));
        assert!(hb.due(HEARTBEAT_MS, Some(0)));
    }

    #[test]
    fn wrap_safe_elapsed() {
        assert_eq!(elapsed(5, 1), 4);
        assert_eq!(elapsed(0, u64::MAX), 1);
        let period = TelemetryInterval::new(TELEMETRY_MIN_MS).unwrap();
        assert!(period.due(TELEMETRY_MIN_MS - 1, Some(u64::MAX)));
    }
}
